src/TableOperations.jl

```diff
@@ -309,9 +311,33 @@ function joinpartitions(x)
             N = length(schema[].names)
             foreach(i -> push!(joined, ChainedVector([Tables.getcolumn(cols, i)])), 1:N)
         else
-            foreach(i -> append!(joined[i], Tables.getcolumn(cols, i)), 1:N)
+            foreach(1:N) do i
+                prev = joined[i]
+                col = Tables.getcolumn(cols, i)
+                T = eltype(prev)
+                S = eltype(col)
+
+                if !(S <: T) && promote
+                    R = promote_type(S, T)
+                    # Promote the joined arrays
+                    newcol = similar(prev, R)
```

Comment:
Since we're using a `ChainedVector` we accept the default (existing length) to avoid changing the underlying blocks.